### PR TITLE
gh-100870: A per-class type cache

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
See https://github.com/faster-cpython/ideas/issues/532.

The main benefit is shrinking the inline cache for `LOAD_ATTR`.

One design flaw is that there is no way to reclaim space occupied by methods that no longer exist (e.g if a method is deleted). My assumption is that most normal code doesn't continuously delete methods though.

* Issue: gh-100870

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
@markshannon could you please run this on the Faster CPython pyperformance machine?

File: Include/internal/pycore_code.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ typedef struct {
     uint16_t counter;
     uint16_t type_version[2];
     uint16_t keys_version[2];
```
Comment (Fidget-Spinner) [inline Include/internal/pycore_code.h:66]: We can shrink this even further by reducing type_version and keys_version in the future.

**Comment by Fidget-Spinner** [comment]
An alternative approach is to store the index into the global type cache instead. That will save us some space since we will no longer need to allocate a per-class cache. We can also make it effectively clean up dead methods id they're all in there.

**Comment by gvanrossum** [comment]
> One design flaw is that there is no way to reclaim space occupied by methods that no longer exist (e.g if a method is deleted). My assumption is that most normal code doesn't continuously delete methods though.

But what about systems that repeatedly reload a module? That doesn't strike me as totally abnormal.

**Comment by Fidget-Spinner** [comment]
I can overcome that with a little more space by attaching the type cache index to the method descriptor. Then during cleanup I set that index to NULL.

**Comment by encukou** [comment]
@Fidget-Spinner, what's the status of this PR?

**Comment by Fidget-Spinner** [comment]
@encukou I think we might need this for free-threaded eventually. But I'd rather open a new PR at that point. So time to close this,

**Pull request closed without merging**